### PR TITLE
[Feature][Fix] Qwen3-TTS V2 Model Runner adaptation and bug fixes

### PR DESCRIPTION
**Pull request by Sy0307**
## Summary

Implement full V2 Model Runner support for Qwen3-TTS (Talker AR + Code2Wav generation stages) and fix critical bugs discovered during migration. After these changes, V2 performance matches V1 within ~5% across all concurrency levels.

### V2 Model Runner adaptation

- Implement preprocess → MTP → forward → postprocess execution chain in `OmniGPUModelRunner.execute_model` for Talker AR stage
- Implement `OmniGenerationModelRunner` execute_model + sample_tokens for Code2Wav non-autoregressive stage (lifecycle, pooler_output construction, dummy EOS for scheduler completion detection)
- Static `inputs_embeds` buffer in `OmniModelState` for FULL CUDA graph compatibility with preprocess
- Static MTP buffers for zero-allocation `torch.cat` per decode step
- `OmniAsyncOutput` for non-blocking D2H copy of pooler_output
- Skip `warmup_kernels` for V2 Omni AR models to prevent state pollution

### Bug fixes

- **MRoPE positions (3D→1D regression)**: V2 passed 1D positions `[N]` instead of 3D M-RoPE positions `[3, N]`. TTS talker_config declares `mrope_section` (`uses_mrope=True`) but the model doesn't implement `SupportsMRoPE`, causing `get_rope_state` to fall back to 1D. Wrong RoPE → wrong attention → model never generates EOS → hits `max_tokens` every time (~25s vs normal ~1s).
  - Fix: `_safe_get_rope` creates `RopeState(num_dims=3)` + monkey-patches `get_mrope_input_positions` when `uses_mrope=True` but model lacks the interface.

- **`OmniModelConfig.eos_token_id`**: Neither `OmniModelConfig` nor upstream `ModelConfig` has `eos_token_id` as a direct attribute.
  - Fix: `getattr` fallback chain (`model_config` → `hf_text_config` → `0`).

- **int32 overflow in Code2Wav batch**: `code_predictor` edge cases produce values like `2^32` (4294967296) in `audio_codes`. V2's `req_states.all_token_ids` uses `torch.int32` → overflow crash in `async_tensor_h2d`.
  - Fix: unified `_sanitize_codec_codes(codes)` clamp on **all** codec entry points (async frame, sync frame filter, ref_code — both paths).

- **async_chunk premature finish**: In async_chunk mode, `update_from_output` marked requests finished after processing only the first chunk (`num_computed_tokens >= len(prompt_token_ids)` was true because `prompt_token_ids` only contained the current chunk).
  - Fix: 3-way condition — `(already_stopped) OR (all_computed AND no_adapter) OR (all_computed AND adapter_done)`.

### Other changes

- `stop_token_ids: [0]` added to Code2Wav `default_sampling_params` in all 3 TTS stage config YAMLs
- Removed unused `_EMPTY` mutable global constant
- Removed no-op `request.stop_reason = request.stop_reason` self-assignment
- `_CODEBOOK_SIZE` promoted to module-level constant (was duplicated as local var)

### Known limitations

- `OmniGenerationModelRunner.profile_run` is a no-op stub (first inference includes cold start)
- `OmniGenerationModelRunner.sample_tokens` does synchronous D2H (no `AsyncOutput`) — acceptable since generation models produce output once per request, not per decode step
- MRoPE fix uses monkey-patch because the model doesn't implement `SupportsMRoPE`; proper fix requires upstream model-side change
- `capture_model` accesses `cudagraph_manager` private attributes (`_capture_descs`, `_candidates`) to exclude FULL mode for tuple-returning models — upstream lacks a public API for this
- Only validated on Qwen3-TTS-12Hz-1.7B-Base; other models untested

## Benchmark results

Rigorous benchmark: warmup 3 rounds + 7 measurement rounds, median reported. Async chunk mode, `temperature=0.9`, `seed=42`.

### Wall time (median, seconds)

| Concurrency | V1 | V2 | V2/V1 |
|-------------|------|------|-------|
| C=1 | 1.147 | 1.153 | 1.00x |
| C=2 | 2.116 | 2.123 | 1.00x |
| C=4 | 2.889 | 3.150 | 1.09x |
| C=8 | 5.239 | 5.446 | 1.04x |

### Per-stage timing (median, ms)

| C | V1 Stage 0 | V1 Stage 1 | V2 Stage 0 | V2 Stage 1 |
|---|-----------|-----------|-----------|-----------|
| 1 | 1092 | 1144 | 1107 | 1158 |
| 2 | 2060 | 2113 | 2069 | 2120 |
| 4 | 2702 | 2886 | 2516 | 3147 |
| 8 | 3810 | 5234 | 3636 | 5440 |

### Per-prompt throughput

| C | V1 | V2 |
|---|------|------|
| 1 | 1.147s | 1.153s |
| 2 | 1.058s | 1.062s |
| 4 | 0.722s | 0.787s |
| 8 | 0.655s | 0.681s |

## Test plan

- [x] `python3 -m pytest tests/core/sched/test_generation_scheduler_finish_condition.py -v` (9 passed)
- [x] `python3 -m pytest tests/worker_v2/test_omni_generation_model_runner.py -v` (10 passed)
- [x] E2E: `VLLM_OMNI_USE_V2_RUNNER=1 python3 end2end.py --query-type Base` (async C=1 )
- [x] E2E batch: V2 C=1,2,4,8 all pass with benchmark (warmup 3 + runs 7)
- [x] V1 regression: V1 C=1,2,4,8 all pass, performance unchanged

cc @tzhouam

### CODE REVIEW

**Review by tzhouam — APPROVED** [review]
LGTM

**Comment by Sy0307** [comment]
### BugFix : Audio quality regression (UTMOS 4.27 → 1.22)

**Root cause:** V2's `OmniGenerationModelRunner` was missing V1's `_update_request_states` logic. In async_chunk mode, the Talker produces codec frames each decode step, and the `ChunkTransferAdapter` replaces `Request.prompt_token_ids` on the Code2Wav side with the new chunk's tokens. V1 handles this by removing the request from `input_batch` and re-adding it, so each chunk is treated as a fresh prefill. V2 inherits upstream `GPUModelRunner`, which treats cached requests as decode steps — so Code2Wav only received `input_ids length=1` (the previous step's EOS token) instead of the full codec frames.

Additionally, V2's `sample_tokens` returned EOS as `sampled_token_ids` every step, causing the scheduler's `check_stop` to immediately mark the request as FINISHED. Subsequent chunks could never be processed.

**Fix:** Added `_handle_async_chunk_updates()`, called in `execute_model` before `add_requests`/`update_requests`. For each cached request with updated `prompt_token_ids`:
- **Phase 1:** Remove stale `req_states` + `model_state` (including intermediate buffer and plugin cleanup)
- **Phase 2:** Re-add with the new chunk's tokens (`num_computed_tokens=0`), routed through `model_state.add_request` with a synthetic `OmniNewRequestData` to ensure `_resolve_additional_information` and plugin notifications run properly

Changed `sampled_token_ids` from `[[eos_id]] * num_reqs` to `[[] for _ in range(num_reqs)]` — empty per-request lists that don't trigger `check_stop`. Request termination is handled by the orchestrator via `chunk_transfer_adapter.finished_requests`.

### Verification

```
========================================================================
Audio Quality Report — V2 Final vs V1 Baseline
========================================================================
V2-FINAL : dur=5.20s  UTMOS=4.3995  PESQ=3.82  STOI=0.9983  SI-SDR=31.3
V1-BASE  : dur=6.00s  UTMOS=4.2697  PESQ=3.60  STOI=0.9975  SI-SDR=31.8
========================================================================
```

PTAK @linyueqian

**Comment by Fattysand** [comment]
**Off-by-one in `in_initial_phase` boundary check**
 
`vllm-omni/vllm_omni/model_executor/stage_input_processors/qwen3_tts.py` ~L170:
 
```python
in_initial_phase = initial_chunk_size > 0 and initial_chunk_size < chunk_size and length < chunk_size
```
 
The condition uses strict `<` (`length < chunk_size`), which causes the initial chunk (IC) phase to end one step too early when `chunk_size` equals `left_context_size`.
 
**Example** (assuming `chunk_size = left_context_size = 25`, `initial_chunk_size = 5`):
 
With the current logic, when `length` reaches 25, `25 < 25` evaluates to `False`, so the IC phase exits prematurely. The actual IC iterations are only: `1–5, 1–10, 1–15, 1–20` (4 steps instead of 5). The first normal-phase chunk then uses `1–45` as context to generate frames `21–45` — meaning the left context is only 20 tokens, and subsequent chunks follow this shifted window (e.g., `21–70` generating `46–70`).
 
**Expected behavior**: IC should complete the full sequence `1–5, 1–10, 1–15, 1–20, 1–25`, so the first normal-phase chunk operates on `1–50` with a proper 25-token left context, generating `26–50`. This keeps the chunk boundaries aligned with what `chunk_size` intends.
 
Should this be `length <= chunk_size` (or equivalently, adjusting the boundary so the last IC step at exactly `chunk_size` is included)?

**Comment by Fattysand** [comment]
`tests/worker_v2/test_init_model_state.py` fails to import because it references `_OMNI_ARCHITECTURES` and `_upstream_init_model_state`, which no longer exist in `vllm_omni.worker_v2.model_states`.
The dispatch logic has moved to `OmniGPUModelRunner.load_model` (inject/restore via `setattr`), so `init_omni_model_state` now unconditionally returns `OmniModelState`.
The three tests relying on the removed symbols need to be deleted. It might be worth adding test coverage for the upstream delegation behavior in `OmniGPUModelRunner.load_model` as well.

**Comment by Sy0307** [comment]
> **Off-by-one in `in_initial_phase` boundary check**
> 
> `vllm-omni/vllm_omni/model_executor/stage_input_processors/qwen3_tts.py` ~L170:
> 
> ```python
> in_initial_phase = initial_chunk_size > 0 and initial_chunk_size < chunk_size and length < chunk_size
> ```
> 
> The condition uses strict `<` (`length < chunk_size`), which causes the initial chunk (IC) phase to end one step too early when `chunk_size` equals `left_context_size`.
> 
> **Example** (assuming `chunk_size = left_context_size = 25`, `initial_chunk_size = 5`):
> 
> With the current logic, when `length` reaches 25, `25 < 25` evaluates to `False`, so the IC phase exits prematurely. The actual IC iterations are only: `1–5, 1–10, 1–15, 1–20` (4 steps instead of 5). The first normal-phase chunk then uses `1–45` as context to generate frames `21–45` — meaning the left context is only 20 tokens, and subsequent chunks follow this shifted window (e.g., `21–70` generating `46–70`).
> 
> **Expected behavior**: IC should complete the full sequence `1–5, 1–10, 1–15, 1–20, 1–25`, so the first normal-phase chunk operates on `1–50` with a proper 25-token left context, generating `26–50`. This keeps the chunk boundaries aligned with what `chunk_size` intends.
> 
> Should this be `length <= chunk_size` (or equivalently, adjusting the boundary so the last IC step at exactly `chunk_size` is included)?

If we change like this, we may also need to change `initial_coverage`. In fact if so we also need to increase an extra initial chunk to be emitted.

**Comment by Sy0307** [comment]
> `tests/worker_v2/test_init_model_state.py` fails to import because it references `_OMNI_ARCHITECTURES` and `_upstream_init_model_state`, which no longer exist in `vllm_omni.worker_v2.model_states`. The dispatch logic has moved to `OmniGPUModelRunner.load_model` (inject/restore via `setattr`), so `init_omni_model_state` now unconditionally returns `OmniModelState`. The three tests relying on the removed symbols need to be deleted. It might be worth adding test coverage for the upstream delegation behavior in `OmniGPUModelRunner.load_model` as well.

Thanks a lot :)

**Comment by Fattysand** [comment]
> > `tests/worker_v2/test_init_model_state.py` fails to import because it references `_OMNI_ARCHITECTURES` and `_upstream_init_model_state`, which no longer exist in `vllm_omni.worker_v2.model_states`. The dispatch logic has moved to `OmniGPUModelRunner.load_model` (inject/restore via `setattr`), so `init_omni_model_state` now unconditionally returns `OmniModelState`. The three tests relying on the removed symbols need to be deleted. It might be worth adding test coverage for the upstream delegation behavior in `OmniGPUModelRunner.load_model` as well.
> 
> Thanks a lot :)

I looked into it further and noticed that `fish_speech.py` already implements the behavior I'm describing. The fix should be self-contained — only two lines in `qwen3_tts.py` need to change.

**`fish_speech.py` (lines 118 & 131):**

    in_initial_phase = initial_chunk_size > 0 and length <= chunk_size
    initial_coverage = (chunk_size // initial_chunk_size) * initial_chunk_size if initial_chunk_size > 0 else 0

**`qwen3_tts.py` (lines 170 & 182–184):**

    in_initial_phase = initial_chunk_size > 0 and initial_chunk_size < chunk_size and length < chunk_size
    initial_coverage = (
        ((chunk_size - 1) // initial_chunk_size) * initial_chunk_size if 0 < initial_chunk_size < chunk_size else 0
    )

The `<` and `chunk_size - 1` are a matched pair — both constrain IC coverage to **strictly less than** `chunk_size`, while `fish_speech.py` uses `<=` and no `-1`, allowing IC to cover **up to** `chunk_size`. This causes `qwen3_tts.py` to miss the last IC chunk (e.g. `cs=25, ic=5`: IC emits at 5,10,15,20 then jumps to normal phase emitting 21–45, skipping a 1–25 emit).

**Proposed fix (align with `fish_speech.py`):**

    # line 170: < → <=
    in_initial_phase = initial_chunk_size > 0 and initial_chunk_size < chunk_size and length <= chunk_size

    # lines 182-184: remove -1
    initial_coverage = (
        (chunk_size // initial_chunk_size) * initial_chunk_size if 0 < initial_chunk_size < chunk_size else 0
    )

**Edge cases verified:**

- **Non-divisible** (`cs=25, ic=8`): `(24//8)*8 == (25//8)*8 == 24`, behavior unchanged.
- **`ic == chunk_size`**: Guarded by `initial_chunk_size < chunk_size`, IC skipped entirely — unaffected.
- **`finished=True` during IC**: Handled by existing `context_length` logic (lines 176–178).

**Comment by Sy0307** [comment]
@tzhouam Please also update `vllm-project:dev/migrate-MR-v2` branch to follow vllm-omni main.
